pkg/apiserver/handler_impersonation.go

```diff
@@ -0,0 +1,172 @@
+				username = impersonationRequest.Name
+
+			case api.Kind("Group"):
+				actingAsAttributes.Resource = "groups"
+				groups = append(groups, impersonationRequest.Name)
+
+			case authenticationapi.Kind("UserExtra"):
+				actingAsAttributes.Resource = "userextras"
+
+			default:
+				glog.V(4).Infof("unknown impersonation request type: %v\n", impersonationRequest)
+				forbidden(w, req)
+				return
+			}
+
+			allowed, reason, err := a.Authorize(actingAsAttributes)
+			if err != nil || !allowed {
+				glog.V(4).Infof("Forbidden: %#v, Reason: %s, Error: %v", req.RequestURI, reason, err)
+				forbidden(w, req)
+				return
+			}
+		}
+
+		newUser := &user.DefaultInfo{
+			Name:   username,
+			Groups: groups,
+			Extra:  userExtra,
+		}
+		requestContextMapper.Update(req, api.WithUser(ctx, newUser))
+
+		oldUser, _ := api.UserFrom(ctx)
+		httplog.LogOf(req, w).Addf("%v is acting as %v", oldUser, newUser)
+
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// buildImpersonationRequests returns a list of objectreferences that represent the different things we're requesting to impersonate.
+// Also includes a map[string][]string representing user.Info.Extra
+// Each request must be authorized against the current user before switching contexts.
+func buildImpersonationRequests(headers http.Header) ([]api.ObjectReference, map[string][]string, error) {
+	impersonationRequests := []api.ObjectReference{}
+	userExtra := map[string][]string{}
+
+	requestedUser := headers.Get(authenticationapi.ImpersonateUserHeader)
+	hasUser := len(requestedUser) > 0
+	if hasUser {
+		if namespace, name, err := serviceaccount.SplitUsername(requestedUser); err == nil {
+			impersonationRequests = append(impersonationRequests, api.ObjectReference{Kind: "ServiceAccount", Namespace: namespace, Name: name})
+		} else {
+			impersonationRequests = append(impersonationRequests, api.ObjectReference{Kind: "User", Name: requestedUser})
+		}
+	}
+
+	hasGroups := false
+	for _, group := range headers[authenticationapi.ImpersonateGroupHeader] {
+		hasGroups = true
+		impersonationRequests = append(impersonationRequests, api.ObjectReference{Kind: "Group", Name: group})
+	}
+
+	hasUserExtra := false
+	for headerName, values := range headers {
+		if !strings.HasPrefix(headerName, authenticationapi.ImpersonateUserExtraHeaderPrefix) {
+			continue
+		}
+
+		hasUserExtra = true
+		extraKey := strings.ToLower(headerName[len(authenticationapi.ImpersonateUserExtraHeaderPrefix):])
+		userExtra[extraKey] = append([]string(nil), values...) // make a copy to ensure we never mutate a value in the original slice by accident
+		impersonationRequests = append(impersonationRequests,
+			api.ObjectReference{
+				Kind: "UserExtra",
+				// we only parse out a group above, but the parsing will fail if there isn't SOME version
+				// using the internal version will help us fail if anyone starts using it
+				APIVersion: authenticationapi.SchemeGroupVersion.String(),
+				Name:       extraKey,
```

Comment:
> Yes, I don't think it's safe to ignore the values. For example the Keystone authenticator / authorization plugins plan to populate role and project fields[0]. I know that PR isn't finalized, but I think it's a good example of how implementations will use these values.

Ok, will do.  I think that structure for subresource=extrakey and name=value plays nicer with the rbac ruleset than the other way around since the value is like to vary more than the key.  Agree?